### PR TITLE
OCPBUGS-22095: Add default MCP objects for rendering

### PR DESCRIPTION
**Pull request by eggfoobar**
added ability to use the default rmaster worker mcp labels for rendering during boostrap

/hold

### CODE REVIEW

File: pkg/performanceprofile/controller/performanceprofile/components/consts.go [hunk 1]
```diff
@@ -12,6 +12,8 @@ const (
 	ComponentNamePrefix = "performance"
 	// MachineConfigRoleLabelKey is the label key to use as label and in MachineConfigSelector of MCP which targets the performance profile
 	MachineConfigRoleLabelKey = "machineconfiguration.openshift.io/role"
+	// MachineConfigPoolRoleLabelPrefix
+	MachineConfigPoolRoleLabelPrefix = "pools.operator.machineconfiguration.openshift.io/"
```
Comment (MarSik) [inline pkg/performanceprofile/controller/performanceprofile/components/consts.go:16]: I would probably move this to the create function. Just to keep it close to the other hardcoded names.

File: pkg/performanceprofile/cmd/render/render.go [hunk 2]
```diff
@@ -159,6 +159,13 @@ func render(inputDir, outputDir string) error {
 		}
 	}
 
+	// If we find no MCPs present in the directory, then we assume the default
+	// Master, Worker ones and make sure that any PAO manifests that specify a Master or Worker label
+	// get correctly rendered. Specifying MCPs outside of the default will still cause a failure as is expected.
+	if len(mcPools) == 0 {
```
Comment (MarSik) [inline pkg/performanceprofile/cmd/render/render.go:163]: There is a corner case here that I am not sure we want address however. An SNO with extra workers might in theory use master + eg. worker-cnf MCPs. The extra worker-cnf will make this condition return false and the master pool will be missing.

**Comment by MarSik** [comment]
@eggfoobar I like the direction, but we should open a bug for this to allow backports and probably make the commit messages a bit nicer.

**Comment by MarSik** [comment]
/retitle OCPBUGS-22095: [WIP] Add default MCP objects for rendering

File: pkg/performanceprofile/controller/performanceprofile/components/consts.go (hunk 1 above)
Comment (eggfoobar) [reply pkg/performanceprofile/controller/performanceprofile/components/consts.go]: Sounds good, done

File: pkg/performanceprofile/cmd/render/render.go [hunk 3]
```diff
@@ -213,6 +213,11 @@ func render(inputDir, outputDir string) error {
 			componentObj.SetOwnerReferences(or)
 		}
 
+		// TODO: Resolve why kubelet ownership breaks bootstrap
+		components.KubeletConfig.SetOwnerReferences(nil)
```
Comment (ffromani) [inline pkg/performanceprofile/cmd/render/render.go:217]: generalizing here: https://github.com/openshift/cluster-node-tuning-operator/pull/907

File: pkg/performanceprofile/cmd/render/render.go (hunk 2 above)
Comment (MarSik) [reply pkg/performanceprofile/cmd/render/render.go:167]: This is the only concern I have, I would rather create only the pools that are missing.

Eg. if no master is loaded from the folder, inject the master here. The same for worker.

File: pkg/performanceprofile/cmd/render/render.go (hunk 3 above)
Comment (eggfoobar) [reply pkg/performanceprofile/cmd/render/render.go:213]: Nice! Should we keep this here for now and remove it in your PR?

File: pkg/performanceprofile/cmd/render/render.go (hunk 2 above)
Comment (eggfoobar) [reply pkg/performanceprofile/cmd/render/render.go]: Ah my mistake, missed this, updated the code to account for this, please give it another look.

File: pkg/performanceprofile/cmd/render/render.go (hunk 3 above)
Comment (ffromani) [reply pkg/performanceprofile/cmd/render/render.go:213]: Either work. If we keep the fix in this PR, I'd avoid setting the OwnerReference in the first place rather than adding (lines 203-214) and then remove it right afterwards.
On top of that, I'm quite persuaded _all_ ownerReferences are wrong because autogenerating the UID is wrong in the first place, `kubeletconfig` is just the first failing here

File: pkg/util/manifests.go [hunk 4]
```diff
@@ -140,3 +140,16 @@ func createLabeledDefaultMCPManifests() []*mcfgv1.MachineConfigPool {
 		},
 	}
 }
+
+func AddGeneratedByAnnotation(annotations map[string]string, profileName, profileNamespace string) map[string]string {
+	const generatedByAnnotationKey = "performanceprofile.openshift.io/generatedby"
```
Comment (ffromani) [inline pkg/util/manifests.go]: generated or rendered?

File: pkg/performanceprofile/cmd/render/render.go (hunk 3 above)
Comment (eggfoobar) [reply pkg/performanceprofile/cmd/render/render.go]: Yeah same, in this latest commit, I'm removing those and triggering an update on ownerref diff to allow performance profile to add them after the fact, from what I can see that doesn't seem to trigger a change on the machine config side. Testing now

File: pkg/performanceprofile/cmd/render/render.go [hunk 5]
```diff
@@ -161,6 +159,9 @@ func render(inputDir, outputDir string) error {
 		}
 	}
 
+	// Append any missing default manifests (i.e. `master`/`worker`)
+	mcPools = util.AppendMissingDefaultMCPManifests(mcPools)
```
Comment (vitus133) [inline pkg/performanceprofile/cmd/render/render.go:163]: Could it be possible to add this functionality to 
https://github.com/openshift/cluster-node-tuning-operator/blob/master/pkg/tuned/cmd/render/render.go
?

File: pkg/performanceprofile/cmd/render/render.go [hunk 6]
```diff
@@ -191,29 +189,12 @@ func render(inputDir, outputDir string) error {
...
-				Kind:       pp.Kind,
-				Name:       pp.Name,
-				APIVersion: pp.APIVersion,
-				UID:        uid,
-			},
-		}
-
 		for _, componentObj := range components.ToObjects() {
-			componentObj.SetOwnerReferences(or)
+			componentObj.SetOwnerReferences(nil)
```
Comment (ffromani) [inline pkg/performanceprofile/cmd/render/render.go:193]: Ideally this logic and this commit (which I think is getting ready to merge) should be split to its own PR because it's a related but different change wrt to "add default MCP objects for rendering"

Comment (eggfoobar) [reply pkg/performanceprofile/cmd/render/render.go]: Sounds good, updated

File: pkg/performanceprofile/cmd/render/render.go (hunk 3 above)
Comment (eggfoobar) [reply pkg/performanceprofile/cmd/render/render.go]: Per our conversation, ownership ref changes will be done in another PR

File: pkg/performanceprofile/cmd/render/render.go (hunk 5 above)
Comment (eggfoobar) [reply pkg/performanceprofile/cmd/render/render.go:165]: That's a good point, the logic seems the same, @jlojosnegros any concerns with adding this call to the tuned renderer?

File: pkg/performanceprofile/cmd/render/render.go [hunk 7]
```diff
@@ -204,6 +207,8 @@ func render(inputDir, outputDir string) error {
 
 		for _, componentObj := range components.ToObjects() {
 			componentObj.SetOwnerReferences(or)
+			anno := util.AddBootstrapRenderedByAnnotation(componentObj.GetAnnotations(), pp.Name, pp.Namespace)
```
Comment (MarSik) [inline pkg/performanceprofile/cmd/render/render.go:210]: I think I would remove this too to make this PR solely about the MCPs.

Comment (eggfoobar) [reply pkg/performanceprofile/cmd/render/render.go]: Sure thing, removed

File: pkg/performanceprofile/cmd/render/render.go (hunk 5 above)
Comment (eggfoobar) [reply pkg/performanceprofile/cmd/render/render.go:165]: Confirmed in Slack, added to tuned renderer

File: pkg/util/manifests.go [hunk 8]
```diff
@@ -97,3 +100,68 @@ func ListFilesFromMultiplePaths(dirPaths []string) ([]string, error) {
 	}
 	return results, nil
 }
+
+// AppendMissingDefaultMCPManifests When default MCPs are missing, it is desirable to still generate the relevant
+// files based off of the standard MCP labels and node selectors.
+//
+// Here we create the default `master` and `worker` MCP if they are missing with their respective base Labels and NodeSelector Labels,
+// this allows any resource such as PAO's to utalize the deault during bootstrap rendering.
```
Comment (yanirq) [inline pkg/util/manifests.go:108]: nits:  s/PAO's/PAO       s/utalize/utilize      s/deault/default

**Comment by vitus133** [comment]
@eggfoobar @jlojosnegros I think we now have a problem in [pkg/tuned/cmd/render/render.go](https://github.com/openshift/cluster-node-tuning-operator/pull/833/files#diff-f3655cb37779a7069442b7a63925fe91be42f16cafb43968dba654f44abf19c5) line 147:

```
if len(filteredPerformanceProfiles) == 0 {
		klog.Warningf("Cannot find any PerformanceProfile applicable to MachineConfigPool %s.", mcpName)
		return fmt.Errorf("Cannot find any PerformanceProfile complying with MachineConfigPool %s.", mcpName)
	}
```
Should probably become
```
if len(filteredPerformanceProfiles) == 0 {
		klog.Infof("Cannot find any PerformanceProfile applicable to MachineConfigPool %s.", mcpName)
		return nil
	}
```

**Comment by eggfoobar** [comment]
> @eggfoobar @jlojosnegros I think we now have a problem in [pkg/tuned/cmd/render/render.go](https://github.com/openshift/cluster-node-tuning-operator/pull/833/files#diff-f3655cb37779a7069442b7a63925fe91be42f16cafb43968dba654f44abf19c5) line 147:
> 
> ```
> if len(filteredPerformanceProfiles) == 0 {
> 		klog.Warningf("Cannot find any PerformanceProfile applicable to MachineConfigPool %s.", mcpName)
> 		return fmt.Errorf("Cannot find any PerformanceProfile complying with MachineConfigPool %s.", mcpName)
> 	}
> ```
> 
> Should probably become
> 
> ```
> if len(filteredPerformanceProfiles) == 0 {
> 		klog.Infof("Cannot find any PerformanceProfile applicable to MachineConfigPool %s.", mcpName)
> 		return nil
> 	}
> ```

That makes sense to me, I'll change it

**Comment by jlojosnegros** [comment]
> @eggfoobar @jlojosnegros I think we now have a problem in [pkg/tuned/cmd/render/render.go](https://github.com/openshift/cluster-node-tuning-operator/pull/833/files#diff-f3655cb37779a7069442b7a63925fe91be42f16cafb43968dba654f44abf19c5) line 147:
> 
> ```
> if len(filteredPerformanceProfiles) == 0 {
> 		klog.Warningf("Cannot find any PerformanceProfile applicable to MachineConfigPool %s.", mcpName)
> 		return fmt.Errorf("Cannot find any PerformanceProfile complying with MachineConfigPool %s.", mcpName)
> 	}
> ```
> 
> Should probably become
> 
> ```
> if len(filteredPerformanceProfiles) == 0 {
> 		klog.Infof("Cannot find any PerformanceProfile applicable to MachineConfigPool %s.", mcpName)
> 		return nil
> 	}
> ```

I am getting late to the party and maybe I have missed something but I do not understand this change. 
If there is no PerformanceProfile is an error so why returning `nil`?

**Comment by vitus133** [comment]
> > @eggfoobar @jlojosnegros I think we now have a problem in [pkg/tuned/cmd/render/render.go](https://github.com/openshift/cluster-node-tuning-operator/pull/833/files#diff-f3655cb37779a7069442b7a63925fe91be42f16cafb43968dba654f44abf19c5) line 147:
> > ```
> > if len(filteredPerformanceProfiles) == 0 {
> > 		klog.Warningf("Cannot find any PerformanceProfile applicable to MachineConfigPool %s.", mcpName)
> > 		return fmt.Errorf("Cannot find any PerformanceProfile complying with MachineConfigPool %s.", mcpName)
> > 	}
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Should probably become
> > ```
> > if len(filteredPerformanceProfiles) == 0 {
> > 		klog.Infof("Cannot find any PerformanceProfile applicable to MachineConfigPool %s.", mcpName)
> > 		return nil
> > 	}
> > ```
> 
> I am getting late to the party and maybe I have missed something but I do not understand this change. If there is no PerformanceProfile is an error so why returning `nil`?

Because this PR adds default MCPs unconditionally. With the code as it was, NOT configuring a performance profile becomes an error.

**Comment by jlojosnegros** [comment]
> > > @eggfoobar @jlojosnegros I think we now have a problem in [pkg/tuned/cmd/render/render.go](https://github.com/openshift/cluster-node-tuning-operator/pull/833/files#diff-f3655cb37779a7069442b7a63925fe91be42f16cafb43968dba654f44abf19c5) line 147:
> > > ```
> > > if len(filteredPerformanceProfiles) == 0 {
> > > 		klog.Warningf("Cannot find any PerformanceProfile applicable to MachineConfigPool %s.", mcpName)
> > > 		return fmt.Errorf("Cannot find any PerformanceProfile complying with MachineConfigPool %s.", mcpName)
> > > 	}
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Should probably become
> > > ```
> > > if len(filteredPerformanceProfiles) == 0 {
> > > 		klog.Infof("Cannot find any PerformanceProfile applicable to MachineConfigPool %s.", mcpName)
> > > 		return nil
> > > 	}
> > > ```
> > 
> > 
> > I am getting late to the party and maybe I have missed something but I do not understand this change. If there is no PerformanceProfile is an error so why returning `nil`?
> 
> Because this PR adds default MCPs unconditionally. With the code as it was, NOT configuring a performance profile becomes an error.

Ok, this PR adds "master" and/or "worker" MCPs if they have not been defined in the input folder but, once that's done, the render process should remain the same, so if not having a PP was an error it should remain an error, right?

**Comment by jlojosnegros** [comment]
> > > > I am getting late to the party and maybe I have missed something but I do not understand this change. If there is no PerformanceProfile is an error so why returning `nil`?
> > > 
> > > 
> > > Because this PR adds default MCPs unconditionally. With the code as it was, NOT configuring a performance profile becomes an error.
> > 
> > 
> > Ok, this PR adds "master" and/or "worker" MCPs if they have not been defined in the input folder but, once that's done, the render process should remain the same, so if not having a PP was an error it should remain an error, right?
> 
> I think not having a PP should be legit, it's a user's choice to configure or not configure a PP. We haven't reached the problem before because we have always configured / not configured PP and MCP simultaneously. But now MCP will be always there. I even think that logging a warning here is too far fetched, Info is probably more appropriate

But you are not rendering anything I mean, the main target of that command is to render a MC from a PP, if you return there you have not render anything so either it is an error or you need to continue

**Comment by MarSik** [comment]
> > I think not having a PP should be legit, it's a user's choice to configure or not configure a PP. We haven't reached the problem before because we have always configured / not configured PP and MCP simultaneously. But now MCP will be always there. I even think that logging a warning here is too far fetched, Info is probably more appropriate
> 
> But you are not rendering anything I mean, the main target of that command is to render a MC from a PP, if you return there you have not render anything so either it is an error or you need to continue

I agree, when no PP is present, this must just be a no-op. And the machine config pool injection is virtual only, right? We never render the MCP to the destination folder, we just use it internally.

**Comment by vitus133** [comment]
> > > > > I am getting late to the party and maybe I have missed something but I do not understand this change. If there is no PerformanceProfile is an error so why returning `nil`?
> > > > 
> > > > 
> > > > Because this PR adds default MCPs unconditionally. With the code as it was, NOT configuring a performance profile becomes an error.
> > > 
> > > 
> > > Ok, this PR adds "master" and/or "worker" MCPs if they have not been defined in the input folder but, once that's done, the render process should remain the same, so if not having a PP was an error it should remain an error, right?
> > 
> > 
> > I think not having a PP should be legit, it's a user's choice to configure or not configure a PP. We haven't reached the problem before because we have always configured / not configured PP and MCP simultaneously. But now MCP will be always there. I even think that logging a warning here is too far fetched, Info is probably more appropriate
> 
> But you are not rendering anything I mean, the main target of that command is to render a MC from a PP, if you return there you have not render anything so either it is an error or you need to continue

The main target of this command is to render PP from an MC. If a user have not provided a PP, we have nothing to render, so this step is not needed. By returning nil in this case we tell bootstrap process to continue peacefully to the next step. No PP - nothing happened, let's do other stuff. If we return error here, bootstrap fails, and it's all because user have not provided PP? Doesn't make sense, right?

**Comment by jlojosnegros** [comment]
> > > > > > I am getting late to the party and maybe I have missed something but I do not understand this change. If there is no PerformanceProfile is an error so why returning `nil`?
> > > > > 
> > > > > 
> > > > > Because this PR adds default MCPs unconditionally. With the code as it was, NOT configuring a performance profile becomes an error.
> > > > 
> > > > 
> > > > Ok, this PR adds "master" and/or "worker" MCPs if they have not been defined in the input folder but, once that's done, the render process should remain the same, so if not having a PP was an error it should remain an error, right?
> > > 
> > > 
> > > I think not having a PP should be legit, it's a user's choice to configure or not configure a PP. We haven't reached the problem before because we have always configured / not configured PP and MCP simultaneously. But now MCP will be always there. I even think that logging a warning here is too far fetched, Info is probably more appropriate
> > 
> > 
> > But you are not rendering anything I mean, the main target of that command is to render a MC from a PP, if you return there you have not render anything so either it is an error or you need to continue
> 
> The main target of this command is to render PP from an MC. If a user have not provided a PP, we have nothing to render, so this step is not needed. By returning nil in this case we tell bootstrap process to continue peacefully to the next step. No PP - nothing happened, let's do other stuff. If we return error here, bootstrap fails, and it's all because user have not provided PP? Doesn't make sense, right?

Not the way other commands are done but ok, if that is what it's needed for bootstrap to work.

@MarSik Nope, default MCPs are just used internally but never rendered.

**Review by ffromani — COMMENTED** [review]
I'm not convinced the better course of action is silently swallow errors and keep going, but this is a larger (and too late) concern. Everything else seems reasonnable.

LGTM.

**Comment by jlojosnegros** [comment]
I will prefer a better error handling than returning "ok" when there is no PP but other than that LGTM.

**Review by MarSik — APPROVED** [review]
/lgtm

**Comment by jmencak** [comment]
> /cc @jmencak
> 
> because it impacts non-perfprof code

Thank you Francesco. I don't have a problem with this change.
/hold cancel
